airbyte-integrations/connectors/source-instagram/source_instagram/streams.py

```diff
@@ -203,7 +205,13 @@ def read_records(
             if not insight_record.get(self.cursor_field):
                 insight_record[self.cursor_field] = insight.get("values")[0]["end_time"]
 
-        yield insight_record
+        complete_records = [insight_record]
+        # if insight_list is empty, we don't want to yield an incomplete record and want to stop syncing this stream gracefully
+        if not insight_list:
+            complete_records = []
+            self.logger.warning(f"No data received for base params {json.dumps(base_params)}. It is time to exit gracefully.")
```

Comment:
```suggestion
            # https://developers.facebook.com/docs/instagram-api/guides/insights/
            # If insights data you are requesting does not exist or is currently unavailable
            # the API will return an empty data set instead of 0 for individual metrics.
            self.logger.warning(f"No data received for base params {json.dumps(base_params)}. Since we can't know whether there is no data or the data is temporarily unavailable, stop syncing so as not to miss temporarily unavailable data.")
```

Reply:
@davydov-d does this describe the situation, would you say? Was working to make it clearer both for the developer and the user what was leading to the early exit. 

If this is indeed the case, couldn't a true negative (the data is not unavailable, but the data doesn't exist/there's not data to report able) for a given stream slice stop us from ever advancing the state because we check it every time?

Reply:
Or i guess since we don't have per-stream state, we'd still be emitting a later state and not going back to re-check this... right 🤔

Reply:
@sherifnada maybe you also have insight here with regards to the fix you were recommending: 
> It feels like the fix might be to gracefully exit syncing that stream so that this behavior (re-sync in a few hours) happens naturally
> i.e: don’t checkpoint state but also don’t fail the sync

Reply:
@erohmensing yes, thanks! I dont think it could stop us advancing the state because all the syncs (3 on cloud + 1 integration test) used to failed on the last slice (i.e. fetching today's data). Once the latest data becomes available (I think it's a matter of hours), the state should advance and the sync should succeed

Reply:
In that case this thread is resolved, still might be nice to fail if the cursor gets too old just in case though